File: day1/day1.cpp
#include <cctype>
#include <fstream>
#include <iostream>

#include <fmt/format.h>

using namespace std;

static size_t captcha(const string& txt, size_t offset)
{
	size_t size = txt.size();
	size_t captcha = 0;
	for (size_t i = 0; i < size; i++)
	{
		if (txt[i] == txt[(i+offset)%size] && isdigit(txt[i]))
		{
			captcha += txt[i]-'0';
		}
	}
	return captcha;
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		cerr << fmt::format("Usage: {} <filename>", argv[0]) << endl;
		return 1;
	}

	ifstream input(argv[1]);
	if (!input)
	{
		cerr << fmt::format("Cannot open {}", argv[1]) << endl;
		return 1;
	}
	string txt;
	getline(input, txt);
	input.close();
	if (input.fail())
	{
		cerr << "Cannot parse the data" << endl;
		return 1;
	}

	cout << fmt::format("Part1: {}", captcha(txt, 1)) << endl
	     << fmt::format("Part2: {}", captcha(txt, txt.size()/2)) << endl;

	return 0;
}

File: day1/day1.py
#!/usr/bin/env python3

def captcha(txt, offset):
    size = len(txt)
    return sum(int(txt[i]) for i in range(size) if txt[i] == txt[(i+offset)%size])

# assert captcha("1122", 1) == 3
# assert captcha("1111", 1) == 4
# assert captcha("1234", 1) == 0
# assert captcha("91212129", 1) == 9

# assert captcha("1212", 2) == 6
# assert captcha("1221", 2) == 0
# assert captcha("123425", 3) == 4
# assert captcha("123123", 3) == 12
# assert captcha("12131415", 4) == 4

import sys
if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read().strip()
    except:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)

    print("Part1:", captcha(txt, 1))
    print("Part2:", captcha(txt, len(txt)//2))
